executor/builder.go

```diff
@@ -1467,7 +1467,30 @@ func (b *executorBuilder) updateForUpdateTSIfNeeded(selectPlan plannercore.Physi
 	if _, ok := selectPlan.(*plannercore.PointGetPlan); ok {
 		return nil
 	}
-	return UpdateForUpdateTS(b.ctx, 0)
+	// The Repeatable Read transaction use Read Committed level to read data for writing (insert, update, delete, select for update),
+	// We should always update the for-update-ts no matter the isolation level is RR or RC.
+	return b.updateForUpdateTS()
+}
+
+// updateForUpdateTS is used to update the for-update-ts for reading data at read consistency level in pessimistic transaction.
+// It could use the cached tso from the statement future to avoid get tso many times.
+func (b *executorBuilder) updateForUpdateTS() error {
+	var newForUpdateTS uint64
+	future, cachedTS := b.ctx.GetSessionVars().TxnCtx.GetStmtFuture()
+	if cachedTS != 0 {
+		newForUpdateTS = cachedTS
+	} else if future != nil {
+		newTS, err := future.Wait()
```

Comment:
Maybe change the `err` to some `waitErr` , since we won't report this error to the caller, and make it easy to know we will fallthrough the error path